proxyclient/experiments/ohmmeter.py

```diff
@@ -0,0 +1,260 @@
+
+    ASP_RX_EN = 0x5020, Register8
+    ASP_TX_EN = 0x5024, Register8
+
+    ASP_RXSLOT_CH1_LSB = 0x5028, Register8
+    ASP_RXSLOT_CH1_MSB = 0x5029, Register8
+
+    ASP_RXSLOT_CH2_LSB = 0x502c, Register8
+    ASP_RXSLOT_CH2_MSB = 0x502d, Register8
+
+    ASP_TXSLOT_CH1_LSB = 0x5068, Register8
+    ASP_TXSLOT_CH1_MSB = 0x5068, Register8
+
+
+class CS42L84:
+    def __init__(self, bus, addr):
+        self.regs = CS42L84Registers(self, 0)
+        self.bus = bus
+        self.addr = addr
+
+    def write(self, regaddr, val, width=8):
+        valbytes = val.to_bytes(width//8, byteorder="little")
+        self.bus.write_reg(self.addr, regaddr, valbytes, regaddrlen=2)
+
+    def read(self, regaddr, width=8):
+        read = self.bus.read_reg(self.addr, regaddr, width//8, regaddrlen=2)
+        return int.from_bytes(read, byteorder='little')
+
+
+def read_devid():
+    pass
+
+def sense_Z():
+    r = l84.regs
+
+    r.MSM_CLAMP_CTRL.set(HS_CLAMP_DISABLE=1)
+    r.DAC_CTRL2.set(PULLDOWN_R=E_PULLDOWN_R.R_1K1OHMS)
+    r.DCID_CTRL1.set(Z_RANGE=E_DCID_Z_RANGE.UNK2)
+    r.DCID_CTRL2.set(GND_SEL=E_DCID_GND_SEL.HS3)
+    r.MSM_BLOCK_EN3.set(DCID_EN=1)
+    r.DCID_CTRL3.set(START=0)
+    r.DCID_CTRL3.set(START=1)
+
+    while not r.DCID_STATUS.reg.DONE:
+        pass
+
+    reading = r.DCID_STATUS.reg.OVERALL
+    offset_trim = r.DCID_TRIM_OFFSET.val - 128
+    slope_trim = r.DCID_TRIM_SLOPE.val - 128
+    pulldown_trim = r.DCID_PULLDOWN_TRIM.val - 128
+
+    Y_overall = ((reading + 0.5) * 0.01086 - 1.0 / (1 - slope_trim * 0.001375)) \
+        / (614.0 + offset_trim * 0.125)
+    Y_pulldown = 1.0 / (1100 - pulldown_trim*2)
+
+    if Y_overall > Y_pulldown:
+        Z_headphones = 1.0 / (Y_overall - Y_pulldown)
+    else:
+        Z_headphones = float('inf')
+
+    r.MSM_BLOCK_EN3.set(DCID_EN=0)
+    r.DAC_CTRL2.set(PULLDOWN_R=E_PULLDOWN_R.NONE)
+
+    return Z_headphones
+
+def init_ring_tip_sense():
+    l84.write(0x1471, 0xe0)
+    l84.write(0x1473, 0xc0)
+    l84.write(0x1477, 0x02)
+    l84.write(0x1478, 0x8f)
+
+    l84.regs.RING_SENSE_CTRL.set(INV=1, UNK1=1,
+        RISETIME=E_DEBOUNCE_TIME.T_125MS, FALLTIME=E_DEBOUNCE_TIME.T_125MS)
+    l84.regs.TIP_SENSE_CTRL.set(INV=1,
+        RISETIME=E_DEBOUNCE_TIME.T_500MS, FALLTIME=E_DEBOUNCE_TIME.T_125MS)
+    l84.regs.MSM_BLOCK_EN3.set(TR_SENSE_EN=1)
+
+def wait_for_plug():
+    while not l84.regs.TR_SENSE_STATUS.reg.TIP_PLUG:
+        time.sleep(0.001)
+
+def wait_for_unplug():
+    while l84.regs.TR_SENSE_STATUS.reg.TIP_UNPLUG:
+        time.sleep(0.001)
+
+
+p.pmgr_adt_clocks_enable("/arm-io/i2c2")
+i2c2 = I2C(u, "/arm-io/i2c2")
+
+p.write32(0x2921f0010, 0x76a02) # invoke reset
+p.write32(0x2921f0010, 0x76a03) # out of reset
+
+l84 = CS42L84(i2c2, 0x4b)
+
+init_ring_tip_sense()
+
+while True:
+    print("Waiting for plug... ", end=""); sys.stdout.flush()
+    wait_for_plug()
+
+    print("measuring... ", end=""); sys.stdout.flush()
+    print(f"{sense_Z():.1f} ohms... ", end=""); sys.stdout.flush()
```

Comment:
You can use the [`flush` keyword argument of `print`](https://docs.python.org/3/library/functions.html#print) to flush stdout instead of calling `sys.stdout.flush()`:

```suggestion
    print("Waiting for plug... ", end="", flush=True)
    wait_for_plug()

    print("measuring... ", end="", flush=True)
    print(f"{sense_Z():.1f} ohms... ", end="", flush=True)
```

Reply:
Ha, forgot `print` can do that, thanks! (Will not be doing a fix right away since `experiments/` don't need to be particularly high quality.)